### ISSUE TITLE
when passing replace = true to mount , exisiting element is not replaced with correct new tag

### ISSUE THREAD

**Issue by devmondo** [leeoniya/domvm#98]
Hi,

maybe this is stupid but i wanted to know if i am doing it wrong or it is a bug or intentional behavior.

place holder to mount the VDom into it

`<div class="placeholder"></div>`

 DOMVM code

```
let domView = domvm.view((vm)=> {
    return ()=>["ul",
        names.map((name)=> {
            return [`li`, name]
        })
    ]
}).mount(document.querySelector(`.placeholder`), true);
```

the result  

```
<div class="placeholder">
    <li>Mike</li>
    <li>Roman</li>
    <li>John</li>
</div>
```

as you see the resulting HTML is having a `DIV` as a root tag instead of `UL`.

the way i think about it that when we chose to replace, then the mount to element should be replaced fully with the correct DOMVM root tag. 

thanks in advanced.

**Comment by leeoniya** [comment]
it's currently assumed that the placeholder element is already the correct type. but i agree that it would make sense to replace it.

**Comment by devmondo** [comment]
thank you :)

**Comment by lawrence-dol** [comment]
> it would make sense to replace it

I would go further and say it's *necessary* to replace it. After all there may be attributes critical to correct behavior defined on the element.

**Comment by leeoniya** [comment]
> I would go further and say it's necessary to replace it.

...except when it's `<body>` ;), which is the primary use-case it was intended for.

**Comment by lawrence-dol** [comment]
> except when it's <body>

Why would body be an exception? It's just as valid to create a body with specific attributes.

But perhaps that's an indicator that the tag should not be changed, but its attributes should be. Personally I think that the example in this question is flawed. A mount-point should be a generic block; nothing other than div or body or main seems to make much sense.

But in terms of expectations, I would expect a "replace" flag to "darn well replace it!"... therefore, if the end result is invalid, that's on the programmer, not DOMVM.

**Comment by leeoniya** [comment]
> I would expect a "replace" flag

that's the thing, it's not a replace flag. it's an isRoot flag right now.

> Why would body be an exception? It's just as valid to create a body with specific attributes.

kind of. you cannot uniformly create and use replaceNode on body (it's special)

http://stackoverflow.com/questions/8010932/how-can-i-add-a-body-element-to-an-empty-dom-document#

**Comment by devmondo** [comment]
@lawrence-dol 
> Personally I think that the example in this question is flawed

that is subjective, the point here it to understand the intended original behavior.

@leeoniya, i did not know it means `isRoot `! but i think the perceived  impression is still the same. if a tag is gonna become a root then it should render the way it was defined.  

this library is really awesome and elegant but may i suggest to give docs more attention as it would really help a lot in clearing some confusion that may occur and increase adaption of the project.

**Comment by leeoniya** [comment]
@devmondo 

> if a tag is gonna become a root then it should render the way it was defined.

agreed

see #36. if you have any interest, help would be appreciated and i'd be happy to clarify/explain any points in there.

**Comment by lawrence-dol** [comment]
> it's an isRoot flag right now

The difference (in my mind) is purely semantic. It's still a root tag that I'd expect to replace the specified dom tree en toto.

> you cannot uniformly create and use replaceNode on body (it's special)

That makes things more difficult. My app works because I don't adorn the body tag with any attributes. But my expectation was that I was specifying the entire body, including the tag and its attributes.

````html
<body>
    <div class="display-panel"><p style="margin:1em; text-align:center;">This application requires JavaScript.<br>If you have JavaScript enabled, please contact Technical Support</p></div>

    <script src="AppScript"></script>

    <script>
    document.addEventListener("DOMContentLoaded", function() {                                      // ensure HTML is fully parsed before replacing body
        new TemJS({
            rootNode                        : document.body,
            // other options here
            });
        });
    </script>
</body>
````

**Comment by devmondo** [comment]
@leeoniya that is a huge list! it would be awesome when it is done 

thanks a lot  for the kind response, this whole VDom is new land for me, i will try to contribute where ever i can :)

**Comment by leeoniya** [comment]
@lawrence-dol there's no roadblock here to getting things working as expected.

for `true`, it will simply recreate & replaceNode the target if the nodeName doesnt match the expected tag *unless* the target is `body`, in which case it will do what it already does (patch attrs, handlers and create the children).

I'll be happy to humor anyone who files a bug expecting a template `div` root to replace `body`.

**Comment by lawrence-dol** [comment]
@leeoniya : Totally agree. I wasn't objecting, just fleshing out the argument from expectations for the new behavior being suggested.

PS: I know I've been dark for a bit, and I will be for a few more weeks. Juggling projects; you know how it goes.

**Comment by lawrence-dol** [comment]
@leeoniya : Totally agree. I wasn't objecting, just fleshing out the argument from expectations for the new behavior being suggested.

PS: I know I've been dark for a bit, and I will be for a few more weeks. Juggling projects; you know how it goes.

**Comment by lawrence-dol** [comment]
And I do have to reiterate, I think the only sensible mounting nodes are body, main, section and div. I personally would prefer to have the view construct a `div` and allow the mount point to determine the extant tag name.

Again, I think the OP expectation of mounting a list is (in the end, not at first look) nonsensical for a mount-over root (that is, the behavior for isRoot). A list is just not a general-purpose container node.

So I guess now, I am arguing against the OP request. If I ask to mount a div onto (as opposed to into) the body, main, or a section, the most useful result is that the block tag in question is retained and its attributes and content are replaced. That allows components to not be dictating to the using code what container element they are to be.

**Comment by devmondo** [comment]
@lawrence-dol  u dint understand why you keep insisting about the example being the intended end result, again the point is to see how the API work.

of course `UL`  example above is not ideal but did you try to mount custom element tag like `carousel` into a `DIV` element? it is still exhibiting the same behavior as above.

and yeah one would use `DIV` as mounting point, but sometimes you have dynamic system and you might not have mounting point and you want to use Technics like `appnedChild`, etc...

**Comment by lawrence-dol** [comment]
> i dont understand why you keep insisting on the example being the intended end result, again the point is to see how the API work.

Don't take that personally. The point is **not only** to understand how the API currently works, but to probe whether it's current behavior is sensible and logical, and explore possible alternative behaviors in the event that it's not.

You now know what DOMVM *currently* does. The question then becomes, *is this what DOMVM should be doing in this circumstance?*

I am proposing that, contrary to my initial thought, which was that when mounting DOM VM should replace the mount-point tag with the tag supplied by the view, it should continue to do what it is currently doing, which is to mount the view on the mount-point tag replacing the tag's attributes and content.

In other words, I've reversed my opinion (for reasons already stated).

**Comment by devmondo** [comment]
@lawrence-dol, nothing is personally taken and any extra opinion by you around this would make it even more helpful.

the big question as i outlined earlier, is how to insert custom elements VDOMs into the page while keeping their properties and Tag. what is a further use case, is that when you want to drag those custom elements and add them to the flow of the page, obviously there is no mount point for any of them. 

i wonder if this sheds more light.

**Comment by leeoniya** [comment]
sorry this took till now. was focused on getting v2 out.

this should be fixed now, plz reopen if it's still an issue.
